src/parser/jobs/blm/modules/Sharpcast.js

```diff
@@ -0,0 +1,136 @@
+
+const SHARPCAST_CONSUMER_IDS = [
+	ACTIONS.FIRE_I.id,
+	ACTIONS.THUNDER_III.id,
+	ACTIONS.THUNDER_IV.id,
+	ACTIONS.SCATHE.id,
+]
+
+export default class Sharpcast extends Module {
+	static handle = 'sharpcast'
+	static dependencies = [
+		'timeline',
+		'procs', // eslint-disable-line @xivanalysis/no-unused-dependencies
+		'suggestions',
+	]
+
+	_buffWindows = {
+		current: null,
+		history: [],
+	}
+
+	_droppedSharpcasts = 0
+
+	constructor(...args) {
+		super(...args)
+		this.addHook('removebuff', {by: 'player', abilityId: STATUSES.SHARPCAST.id}, this._onRemoveSharpcast)
+		this.addHook('applybuff', {by: 'player', abilityId: STATUSES.SHARPCAST.id}, this._onGainSharpcast)
+		this.addHook('cast', {by: 'player', abilityId: SHARPCAST_CONSUMER_IDS}, this._onCast)
+		this.addHook('death', {to: 'player'}, this._onDeath)
+		this.addHook('complete', this._onComplete)
+	}
+
+	_onRemoveSharpcast(event) {
+		this._stopAndSave(event.timestamp)
+	}
+
+	_onGainSharpcast(event) {
+		this._buffWindows.current = {
+			start: event.timestamp,
+		}
+	}
+
+	// Consolidate old onCast functions into one central function
+	_onCast(event) {
+		const actionId = event.ability.guid
+
+		const action = getDataBy(ACTIONS, 'id', actionId)
+		if (action && action.onGcd) {
+			this._tryConsumeSharpcast(event)
+		}
+	}
+
+	_tryConsumeSharpcast(event) {
+		const actionId = event.ability.guid
+
+		// If this action isn't affected by a proc (or something is wrong), bail out
+		if (!SHARPCAST_CONSUMER_IDS.includes(actionId)) {
+			return
+		}
+
+		// If this proc is active, consume it
+		if (this._buffWindows.current) {
+			// Stop the buff window, and ensure it's not marked as a drop
+			this._stopAndSave(event.timestamp, false)
+		}
+	}
+
+	_onDeath(event) {
+		this._stopAndSave(event.timestamp)
+	}
+
+	_stopAndSave(endTime = this.parser.currentTimestamp, countDrops = true) {
+		if (!this._buffWindows.current) {
+			return
+		}
+
+		this._buffWindows.current.stop = endTime
+		if (this._buffWindows.current.stop - this._buffWindows.current.start >= SHARPCAST_DURATION_MILLIS && countDrops) {
+			this._droppedSharpcasts++
+		}
+		this._buffWindows.history.push(this._buffWindows.current)
+		this._buffWindows.current = null
+	}
+
+	_onComplete() {
+		// Finalise the buff if it was still active
+		if (this._buffWindows.current) {
+			this._stopAndSave()
+		}
+
+		const groupId = 'procbuffs-' + STATUSES.SHARPCAST.id
+		const fightStart = this.parser.fight.start_time
+
+		// Make sure a timeline group exists for this buff
+		// Add buff windows to the timeline
+		this._buffWindows.history.forEach(window => {
+			this.timeline.addItem(new Item({
+				type: 'background',
+				start: window.start - fightStart,
+				end: window.stop - fightStart,
+				group: groupId,
+				content: <img src={STATUSES.SHARPCAST.icon} alt={STATUSES.SHARPCAST.name}/>,
+			}))
+		})
+
+		// Suggestions to use sharpcasts that wore off.
+		if (this._droppedSharpcasts > 0) {
```

Comment:
Minor, but you don't need to bother with this check on tiered suggestions; it's handled implicitly.

Reply:
Yeah, I noticed that while working on the dnc module earlier, I'll pull those